add optional support for binary data in database search

GregorLohaus opened the pull request:
Signed-off-by: Gregor Lohaus <lohausgregor@gmail.com>

### Description

Recently I had to search for some binary data in a database, globally. 
Unfortunately neither LIKE nor REGEXP work with hex literals, so i added an option to include matches on hexed column data in the database search module. 
I think some people could benefit from this.

kamil-tekiela commented on src/Database/Search.php:
```diff
@@ -44,6 +44,13 @@ class Search
+     *
+     * @var mixed[]
+     */
+    private array $criteriaSearchOptions;
```
Why is it an array of mixed values if you have only a single boolean?

kamil-tekiela commented on src/Database/Search.php:
```diff
@@ -116,6 +124,12 @@ private function setSearchParams(): void
             $this->criteriaColumnName = $_POST['criteriaColumnName'];
         }
+
+        if (isset($_POST['criteriaSearchOptionIncludeHex']) === false) {
+            $this->criteriaSearchOptions['include_hex_matches'] = false;
+        } else {
+            $this->criteriaSearchOptions['include_hex_matches'] = true;
+        }
```
```suggestion
        $this->criteriaSearchOptions['include_hex_matches'] = isset($_POST['criteriaSearchOptionIncludeHex']);
```

GregorLohaus replied:
the value will now always be set to either '1' or '0' due to the added hidden input, i still have to check for isset because on the first page render the value is not set the and the empty check used for searchType is disallowed by phpstan config

kamil-tekiela commented on src/Database/Search.php:
```diff
@@ -189,6 +203,14 @@ private function getWhereClause(string $table): string
                     . ' ' . $likeOrRegex . ' '
                     . $this->dbi->quoteString($automaticWildcard . $searchWord . $automaticWildcard);
+
+                if (! ((bool) $this->criteriaSearchOptions['include_hex_matches'])) {
```
The cast is unnecessary since the only possible value is a boolean.

GregorLohaus replied:
i think i did this to satisfy phpstan when the value could be unset but i removed the cast now

kamil-tekiela commented on tests/unit/Database/SearchTest.php:
```diff
@@ -88,40 +89,93 @@ public function testGetWhereClause(string $type, string $expected): void
             ],
             [
                 '5',
+                ['include_hex_matches' => null],
```
I see no reason to test `null`. Your condition checks if the value is set, so `['include_hex_matches' => null],` is unnecessary.

GregorLohaus replied:
this was also done to satisfy phpstan

gregor-lohaus committed:
add optional support for binary data in database search

Signed-off-by: gregor <gregor.lohaus@8mylez.com>

GregorLohaus commented:
@kamil-tekiela i have refactored this a bit adding a hidden input so the arraykey will always be set in the post request, i thought using an array might be a good idea if someone wanted to add more options in the future but i have changed this to use a boolean property for now

MauricioFauth commented:
Merged. Thank you!
